### PR TITLE
[Serializer] Add Support for `object_to_populate` in CustomNormalizer

### PR DESCRIPTION
**Pull request by chrisguitarguy**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21715
| License       | MIT
| Doc PR        | n/a

This pulls a trait out of `AbstractNormalizer` with a method to extract the object to populate and adds some tests for it. Then uses that trait in both `AbstractNormalizer` and `CustomNormalizer` so both can support the `object_to_populate` key.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 1]
```diff
@@ -316,12 +318,7 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
         }
 
-        if (
-            isset($context[static::OBJECT_TO_POPULATE]) &&
-            is_object($context[static::OBJECT_TO_POPULATE]) &&
-            $context[static::OBJECT_TO_POPULATE] instanceof $class
-        ) {
-            $object = $context[static::OBJECT_TO_POPULATE];
+        if (null !== $object = $this->extractObjectToPopulate($class, $context, static::OBJECT_TO_POPULATE)) {
             unset($context[static::OBJECT_TO_POPULATE]);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Shouldn't this line be moved in in the trait too?

File: src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+        if (
+            isset($context[$key]) &&
+            is_object($context[$key]) &&
+            $context[$key] instanceof $class
+        ) {
+            return $context[$key];
+        }
+
+        return null;
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php]: Useless, can be removed.

File: src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+trait ObjectToPopulateTrait
+{
+    /**
+     * Extract the `object_to_populate` field from the context if it exists
+     * and is an instance of the provided $class.
+     *
+     * @param string $class The class the object should be
+     * @param $context The denormalization context
+     * @param string $key They in which to look for the object to populate.
+     *                    Keeps backwards compatibility with `AbstractNormalizer`.
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php]: There is no reason to mention this in the comment IMO.

**Review by dunglas — CHANGES_REQUESTED** [review]

**Comment by dunglas** [comment]
Just asking, what is the use case of this class? I think that it's always better to create custom normalizers (implementing the `NormalizerInterface` than adding the serialization logic in the class itself.

I would be in favor of deprecating this class and remove it from Symfony 3.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 1 above)
Comment (chrisguitarguy) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Would involve passing context by reference into the trait. That's the only reason I didn't do it.

**Comment by chrisguitarguy** [comment]
> Just asking, what is the use case of this class?

In my case, I was interested in using it with a "dumb" dto-like object.

```php
class SomeDto implements DenormalizableInterface
{
  // ...
}
```

The idea being to keep everything close together: the normalization, denormalization, and the DTO itself. So basically what you are against :)

I was interested in doing this because I think it's a lot nicer developer experience to keep the normalization/denormalization closer to the object being serialize/deserialized. Easier to understand and catch at a glance. Doesn't require you to register services in the container with a special tag, etc.

File: src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php (hunk 3 above)
Comment (felipsmartins) [inline src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php]: Why?

**Review by dunglas — APPROVED** [review]

**Comment by fabpot** [comment]
@chrisguitarguy Can you please rebase it so that we can merge it? Thanks. Base should also be changed to 3.4.

**Comment by chrisguitarguy** [comment]
@fabpot rebased on 3.4. Did you want me to make any of the changes @dunglas suggested?

**Comment by fabpot** [comment]
Looks like he approved this PR without any further changes :)

**Comment by fabpot** [comment]
Thank you @chrisguitarguy.

**Pull request closed without merging**